compiler/il/OMRNode.cpp

```diff
@@ -2357,7 +2357,103 @@ OMR::Node::isNotCollected()
+ *    i.e. aiadd, aiuadd, aladd, aluadd. Suppport for other opcodes can be added if necessary.
+ */
+bool
+OMR::Node::computeIsInternalPointer()
+   {
+   TR_ASSERT(self()->getOpCode().hasPinningArrayPointer(), "Opcode %s is not supported, node is " POINTER_PRINTF_FORMAT, self()->getOpCode().getName(), self());
+   return self()->computeIsCollectedReference();
+   }
+
+/**
+ * \brief
+ *    Check if the node is collected.
+ *
+ * \return
+ *    Return true if the node is collected, false otherwise.
+ *
+ * \note
+ *    This query traverses down the children of the node and try to find if the node originates
+ *    from something collected, i.e. anything that is collected reference or internal pointer.
+ */
+bool
+OMR::Node::computeIsCollectedReference()
+   {
+   TR::Node *curNode = self();
+   if (curNode->getOpCode().isTreeTop())
+      return false;
+
+   while (curNode)
+      {
+      if (!curNode->getType().isAddress())
+         return false;
 
+      if (curNode->isInternalPointer())
+         return true;
+
+      TR::ILOpCode op = curNode->getOpCode();
+      TR::ILOpCodes opValue = curNode->getOpCodeValue();
+
+      // If a language can handle a collected reference going via a non-address type,
+      // then the logic would need to be augmented to handle that
+      if (op.isConversion())
+         return false;
+
+      // The following are all opcodes that are address type, non-TreeTop and non-conversion
+      switch (opValue)
+         {
+         case TR::aiadd:
+         case TR::aiuadd:
+         case TR::aladd:
+         case TR::aluadd:
+            curNode = curNode->getFirstChild();
+            break;
+         // Symbols for calls and news does not contain collectedness information.
+         // Current implementation treats all object references collectable, and also
+         // assumes that the return of an acall* is an object reference.
+         case TR::acall:
+         case TR::acalli:
+         case TR::New:
+         case TR::newarray:
+         case TR::anewarray:
+         case TR::variableNew:
+         case TR::variableNewArray:
+         case TR::multianewarray:
+              return true;
+         // aload, aloadi, aRegLoad and loadaddr are associated with a symref, we should
+         // be able to tell its collectedness from its symref.
+         case TR::aRegLoad:
+         case TR::aloadi:
+         case TR::aload:
+         case TR::loadaddr:
+            {
+            TR::Symbol *symbol = curNode->getSymbolReference()->getSymbol();
+            // isCollectedReference() responds false to generic int shadows because their type
+            // is int. However, address type generic int shadows refer to collected slots.
+            if (opValue == TR::aloadi && symbol == TR::comp()->getSymRefTab()->findGenericIntShadowSymbol())
+               return true;
+            else
+               return symbol->isCollectedReference();
```

Comment:
It should be ture. Internal pointer temps are address type, and that query is equivalent to `return (self()->getDataType()==TR::Address || self()->isLocalObject()) && !self()->isNotCollected();`